### PR TITLE
Image collection tilesets

### PR DESCRIPTION
**Pull request by DavidMikeSimon**
This PR adds support for image collection tilesets, i.e. tilesets that assign an individual image file to each tile, for use in object layers.

image

I recommend reviewing with the "hide whitespace changes" option enabled, since there's some re-indentation of blocks that makes some parts seem like bigger changes than they really are.

I'm pretty unfamiliar with Bevy, so please let me know if there's something I'm missing, or something I could be doing in a more idiomatic way!

There are still some unresolved issues:
* There is z-fighting between objects that have the same position, even if they are on different layers. To see what I mean, enable the other background layers in `images-map.tmx` and run the `images_main` example again.
* Loading the individual images seems a little slower than I'd expect.
* Before an individual tile image is loaded, sometimes the object will be rendered as an unrelated tile (!!), if the map has both regular tile layers and object layers.
* I haven't even tried object scaling or rotation yet.
* When I add debugging statements in `try_from_bytes`, I notice that Bevy seems to be repeatedly re-reading the map, even though it has not changed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,6 +33,15 @@ cargo run --example iso_main
...
 In these examples, you should be able to use the wasd keys to pan across the maps. You can follow a similar pattern in your own Bevy project. For more information, follow the [Bevy Setup] guide.
+### Example Assets
+
+Assets for the examples are from the following sources:
+
+[Outdoor Tiles Again](https://opengameart.org/content/outdoor-tiles-again) - Michele "Buch" Bucelli
+[Isometric 64x64 Outside Tileset](https://opengameart.org/content/isometric-64x64-outside-tileset) - Yar
+[Jumper Pack](https://kenney.nl/assets/jumper-pack) - Kenney
+
+These are all licensed under variants of the Creative Commons license. Please see each individual site about how to properly credit the creator if you decide to use these assets in your own project.
```
Comment (DavidMikeSimon) [inline README.md:44]: Added some credits for the assets that were already in the repo, in addition to the Kenney assets added in this PR. I found the `opengameart.org` links via Google image search, and they seem legit, but it's hard to tell if it's the true original source.

File: src/map.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ pub struct Map {
     pub map: tiled::Map,
     pub meshes: Vec<(u32, u32, Mesh)>,
     pub layers: Vec<MapLayer>,
-    pub groups: Vec<ObjectGroup>,
+    pub object_groups: Vec<ObjectGroup>,
```
Comment (DavidMikeSimon) [inline src/map.rs:29]: Renamed `groups` to `object_groups` to avoid confusion with TMX `<group>`s, which are distinct from TMX `<objectgroup>`s.

File: src/map.rs [hunk 3]
```diff
@@ -86,23 +86,34 @@ impl Map {
...
         }
 
         let tile_size = Vec2::new(map.tile_width as f32, map.tile_height as f32);
         let image_folder: PathBuf = asset_path.parent().unwrap().into();
         let mut asset_dependencies = Vec::new();
 
+        for tileset in map.tilesets.iter() {
+            for image in tileset.images.iter() {
+                let tile_path = image_folder.join(image.source.as_str());
+                asset_dependencies.push(tile_path);
```
Comment (DavidMikeSimon) [inline src/map.rs:99]: Moved this existing tileset image asset checker here, out of the layers loop, to avoid repeatedly adding the same entry to `asset_dependencies` for each layer.

File: src/map.rs [hunk 4]
```diff
@@ -86,23 +86,34 @@ impl Map {
...
+            }
+
+            for tile in tileset.tiles.iter() {
+                if let Some(tile_image) = tile.images.first() {
+                    let tile_path = image_folder.join(tile_image.source.as_str());
+                    asset_dependencies.push(tile_path);
```
Comment (DavidMikeSimon) [inline src/map.rs:105]: And here we also check for tile-specific individual images.

File: src/map.rs [hunk 5]
```diff
@@ -241,58 +252,89 @@ pub fn process_loaded_tile_maps(
...
+            let object_gids: Vec<_> = map
+                .object_groups
+                .iter()
+                .flat_map(|og| og.objects.iter().map(|o| o.tileset_gid))
+                .collect();
 
             for tileset in &map.map.tilesets {
-                if !materials_map.contains_key(&tileset.first_gid) {
+
+                if materials_map.contains_key(&tileset.first_gid) { continue; }
```
Comment (DavidMikeSimon) [inline src/map.rs:265]: Added some early breaks/continues here to reduce the amount of stair-stepping indentation. Please let me know if you want to revert these changes, they're not semantically important, but I think it's a little easier to read this way.

File: src/map.rs [hunk 6]
```diff
@@ -241,58 +252,89 @@ pub fn process_loaded_tile_maps(
...
-            if map_handle != changed_map {
-                continue;
-            }
+            if map_handle != changed_map { continue; }
+
+            let object_gids: Vec<_> = map
+                .object_groups
+                .iter()
+                .flat_map(|og| og.objects.iter().map(|o| o.tileset_gid))
+                .collect();
```
Comment (DavidMikeSimon) [inline src/map.rs:261]: I moved the generation of `object_gids` out of the tileset loop, since it's the same list for every tileset.

File: src/map.rs [hunk 7]
```diff
@@ -241,58 +252,89 @@ pub fn process_loaded_tile_maps(
...
+                    let atlas_handle = texture_atlases.add(atlas);
+                    for i in 0..(columns * rows) as u32 {
+                        if texture_atlas_map.contains_key(&(tileset.first_gid + i)) {
+                            continue;
+                        }
+                        // println!("insert: {}", tileset.first_gid + i);
+                        texture_atlas_map
+                            .insert(tileset.first_gid + i, atlas_handle.clone());
+                    }
+                } else if let Some(_) = tileset.tiles.first().and_then(|tile| tile.images.first()) {
```
Comment (DavidMikeSimon) [inline src/map.rs:310]: If there's no image on the tileset, here we next check if the first tile has an individual image.

I'm assuming here that either *every* tile in a tileset will have an individual image, or *none* of them will. I think that's right, but maybe there's a more complex Tiled arrangement I'm not aware of?

File: src/map.rs [hunk 8]
```diff
@@ -241,58 +252,89 @@ pub fn process_loaded_tile_maps(
...
+                            materials.add(texture_handle.clone().into())
+                        );
+
+                        // Create an atlas with only this single image
+                        let atlas = TextureAtlas::from_grid(
+                            texture_handle.clone(),
+                            Vec2::new(tile_image.width as f32, tile_image.height as f32),
+                            1,
+                            1
+                        );
```
Comment (DavidMikeSimon) [inline src/map.rs:335]: To reduce the amount of churn needed outside of `map.rs`, we just create a one-image `TextureAtlas` for each individual tile image.

An alternate approach would be to use `TextureAtlasBuilder`, however I'm not sure about the performance implications of that (e.g. is it something that should be happening in a build process outside of the main game executable? does it actually make rendering faster? does it affect loading times?). That could be explored in a separate PR.

File: src/map.rs [hunk 9]
```diff
@@ -323,7 +365,7 @@ pub fn process_loaded_tile_maps(
         mut created_entities,
     ) in query.iter_mut()
     {
-        if new_meshes.contains_key(map_handle) {
+        if changed_maps.contains(map_handle) {
```
Comment (DavidMikeSimon) [inline src/map.rs:368]: Can no longer treat a new mesh as equivalent to an updated map, since individual tile images could change without needing a mesh to be rebuilt.

File: src/map.rs [hunk 10]
```diff
@@ -426,12 +470,17 @@ pub fn process_loaded_tile_maps(
...
+
+                    let atlas_sprite_index = &map.map.tilesets.iter()
+                        .find(|tileset| Some(tileset.first_gid) == object.tileset_gid)
+                        .and_then(|tileset| if tileset.images.len() > 0 { object.sprite_index } else { Some(0) } );
 
                     let entity = object
                         .spawn(
                             &mut commands,
                             atlas_handle,
+                            *atlas_sprite_index,
```
Comment (DavidMikeSimon) [inline src/map.rs:483]: Since some tiles now refer to single-image `TextureAtlas`es, we can no longer assume that TMX sprite index is equivalent to the index in the texture atlas. So, I've added another argument to `spawn` to pass in the atlas index. This may also make it easier to do other kinds of optimizations later? (E.g. splitting or merging tilemaps?)

File: src/objects.rs [hunk 11]
```diff
@@ -144,34 +145,40 @@ impl Object {
...
-                Some((dims / size).extend(1.0))
+            let tile_scale = if tileset.map(|tileset| tileset.images.len() > 0) == Some(true) {
+                if let (Some(dims), Some(size)) = (dims, object_tile_size) {
+                    Some((dims / size).extend(1.0))
+                } else {
+                    None
+                }
             } else {
-                None
+                Some(Vec3::ONE)
```
Comment (DavidMikeSimon) [inline src/objects.rs:175]: This provides the correct positioning and scaling for individual image tiles, since their sizing has nothing to do with the tileset's grid.

File: src/map.rs (hunk 7 above)
Comment (bjorn) [reply src/map.rs:310]: No, that should be a safe assumption.

File: src/map.rs (hunk 8 above)
Comment (bjorn) [reply src/map.rs:335]: Right, it could definitely be interesting to dynamically build the atlas, but indeed some people will be doing that in a build process, which case you'd want to look up the atlas + index by the tile image file name. Probably worth providing both options.

**Comment by DavidMikeSimon** [comment]
@bjorn How's this look to you? Anything you'd like me to change to make it more merge-able?

**Comment by bjorn** [comment]
@DavidMikeSimon Sorry I still know too little about Rust and Bevy to judge how merge-able this change is. I do hope to improve on that with time. :-)

**Comment by DavidMikeSimon** [comment]
@bjorn No worries, and thank you for the helpful comments!

@dmtaub Heya, are you the right person to ask about next steps for this PR?

**Comment by DavidMikeSimon** [comment]
@StarArawn Any feedback on this?

**Comment by sharpner** [comment]
I would love for this to get merged, any chance that is happening? :-)

**Comment by StarArawn** [comment]
Due to time I've pretty much abandoned this project. If someone would like to be the new maintainer feel free to DM me on discord my user name is `StarToaster`. As of bevy 0.6 this library will no longer work, and I don't have any plans to update or upgrade it at the moment. If I had the time I would do the following:

1. Remove the existing `bevy_tiled` renderer.
2. Re-write all of the code to use `bevy_ecs_tilemap`, and expose bevy_ecs_tilemap types.

There is an example of how to render a tiled map using `bevy_ecs_tilemap` which can be found here:
https://github.com/StarArawn/bevy_ecs_tilemap/tree/main/examples/tiled

Why switch to using `bevy_ecs_tilemap`?
It has support for square, hex, and complex isometric tile maps. This library only has square and basic isometric support. Also the renderer is quite robust and is already working with bevy 0.6.

**Comment by sharpner** [comment]
thanks for the update and the heads up!
